### PR TITLE
:hammer: Evaluate search performance for articles

### PR DESCRIPTION
**Pull request by larsyencken**
This PR introduces `make bench.search`, and a search evaluation script.

## Overview

It currently fetches a dataset of synthetic queries and evaluates the extent to which we surface good articles for the given queries.

The scoring algorithm chosen for articles is `precision@4`, meaning the the proportion of the first four results that are relevant, averaged over a ton of queries. The best possible result is 1, the worse is 0.

This is chosen since at most four articles are presented un-collapsed, and the value of getting those four right is much much higher than getting any right further down in the ranking.

It does not yet score chart or explorer search. When we do that, we may move to a more holistic search ranking score like Mean-Average Precision (MAP).

## Current results

```json
{
  "name": "synthetic-queries-2024-03-25.json",
  "scope": "articles",
  "meanPrecision": 0.257,
  "numQueries": 4260,
  "algoliaApp": "74GKBOIDJQ",
  "algoliaIndex": "search-evaluation-algolia-pages"
}
```

## Query datasets

- Single word: https://pub-ec761fe0df554b02bc605610f3296000.r2.dev/synthetic-queries-single-2024-03-25.json
- Single and multi-word: https://pub-ec761fe0df554b02bc605610f3296000.r2.dev/synthetic-queries-2024-03-25.json

### CODE REVIEW

**Comment by larsyencken** [comment]
Following discussion of the instant search results on slack, it now reports `precision@2` and `precision@4` both.

```json
{
  "name": "synthetic-queries-2024-03-25.json",
  "scope": "articles",
  "scores": {
    "precision@2": 0.338,
    "precision@4": 0.257
  },
  "numQueries": 4260,
  "algoliaApp": "74GKBOIDJQ",
  "algoliaIndex": "search-evaluation-algolia-pages"
}
```

File: site/search/evaluateSearch.ts
Comment (marcelgerber) [inline site/search/evaluateSearch.ts:1]: Nice! I'm wondering if it would be helpful to have a verbose output (to a JSON file), that enumerates all the searches and the good/bad results.
Could be helpful to get an overview and find some low-hanging fruits for improvements.

File: site/search/evaluateSearch.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            await next()
+        }
+    }
+
+    const promises = []
+    while (activeQueries < CONCURRENT_QUERIES && i < queries.length) {
+        promises.push(next())
+    }
+
+    await Promise.all(promises)
```
Comment (marcelgerber) [inline site/search/evaluateSearch.ts:166]: Hm, this code is a bit hard to follow in my mind.
Haven't tried the below, but I think we can make it nicer using the `p-map` dependency we're using.


```suggestion
    scores = await pMap(queries, (query) => simulateQuery(index, query), { concurrency: CONCURRENT_QUERIES })
```

File: site/search/evaluateSearch.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ * Simulate searches against our Algolia index and evaluate the results.
+ */
+
+import {
+    ALGOLIA_ID,
+    ALGOLIA_SEARCH_KEY,
+} from "../../settings/clientSettings.js"
+import { SEARCH_EVAL_URL } from "../../settings/serverSettings.js"
+import { getIndexName } from "./searchClient.js"
+import algoliasearch from "algoliasearch"
```
Comment (marcelgerber) [inline site/search/evaluateSearch.ts:11]:
```suggestion
import algoliasearch, { SearchClient, SearchIndex } from "algoliasearch"
```

File: site/search/evaluateSearch.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        name: ds.name,
+        scope: "articles",
+        scores: scores,
+        numQueries: ds.queries.length,
+        algoliaApp: ALGOLIA_ID,
+        algoliaIndex: indexName,
+    }
+}
+
+const getClient = (): any => {
```
Comment (marcelgerber) [inline site/search/evaluateSearch.ts:93]:
```suggestion
const getClient = (): SearchClient => {
```

File: site/search/evaluateSearch.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+const fetchQueryDataset = async (name: string): Promise<QueryDataset> => {
+    const url: string = `${SEARCH_EVAL_URL}/${name}`
+    const resp = await fetch(url)
+    const jsonData = await resp.json()
+    return { name, queries: jsonData }
+}
+
+const simulateQuery = async (
+    index: any,
```
Comment (marcelgerber) [inline site/search/evaluateSearch.ts:106]:
```suggestion
    index: SearchIndex,
```

File: site/search/evaluateSearch.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    query: Query
+): Promise<ScoredQuery> => {
+    const { hits } = await index.search(query.query)
```
Comment (marcelgerber) [inline site/search/evaluateSearch.ts:109]: Just a suggestion, but since we only ever look at the first 4 results anyhow:

```suggestion
    const { hits } = await index.search(query.query, {
        hitsPerPage: N_ARTICLES_LONG_RESULTS,
    })
```

File: site/search/evaluateSearch.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        const correct = actualTruncated.filter((a) =>
+            relevant.includes(a)
+        ).length
+        scores[key] = correct / n
+    }
+    return scores
+}
+
+const simulateQueries = async (
+    index: any,
```
Comment (marcelgerber) [inline site/search/evaluateSearch.ts:136]:
```suggestion
    index: SearchIndex,
```

**Review by marcelgerber — APPROVED** [review]
Brilliant work!
This works very well, and is super fast also.

Some of my review comments are only about using the Algolia types `SearchClient` and `SearchIndex`...

**Comment by marcelgerber** [comment]
Some non-important remarks on the synthetic queries, and our way of scoring them:

We should be aware that it will be impossible to score a precision of 100%.
For example, for the search query `21st`, the synthetic file expects to _only_ return `military-long-run-spending-perspective`, which is unrealistic for this query.
This is - at least partially - a result of the query-prefix generating we're doing as part of the processing. Would be interesting to see the results just for the full-length queries.

Also, another slight flaw of the precision metric is that if for a query, 4 results are expected but we only return a single one (that is among them), then we achieve a score of 100%. But I guess that's just inherent to a precision metric - and we would need to implement recall also if we want to overcome this flaw.

**Comment by marcelgerber** [comment]
@larsyencken I think it makes sense to merge this PR in the current state.
The tool, as it is, is really good and helpful already.

**Comment by marcelgerber** [comment]
I'm just gonna merge this as-is now, seeing as it is already useful, and it doesn't touch any existing code.

**Comment by larsyencken** [comment]
Thanks Marcel!
